File: Snake/elements/Rat.java
package elements;

import java.awt.Color;
import java.awt.Graphics;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Random;

import javax.swing.Timer;

import game.Board;
import game.GameDataDisplay;

public class Rat implements ActionListener{	
	private Cell[][] board;
	private GameDataDisplay display;
	
	private Cell position;
	
	private Timer timer;
	private Random rn;
	private int numberOfRats;
	private int count;	
	
	private static final int SECOND = 1000;
	public static final int DESAPIRING_SECONDS = 10;
	
	public Rat(Board board, GameDataDisplay display){
		super();
		this.display = display;
		this.board = board.getBoard();
		rn = new Random();
		
		setRandomRat();
	}
		
	public void setRandomRat(){		
		int x ;
		int y ;
		
		do{
		x = rn.nextInt(this.board.length);
		y = rn.nextInt(this.board[0].length);
		}while(!(board[x][y].isGrass()));
		
		if(rn.nextInt(101) > 95){
			timer = new Timer(SECOND, this);		
			count = DESAPIRING_SECONDS;
			numberOfRats = rn.nextInt(5) + 1;			
			display.tempRats.setText("" + count);
			display.tempRats.setVisible(true);
			timer.start();
		}
		
		position = board[x][y];
		position.setFigure(new RatFigure());		
	}
	
	public int getNumberOfRats(){return numberOfRats;}
	
	public boolean isTimed(){return timer != null;}
	
	public Cell getPosition(){return position;}
	
	public void actionPerformed(ActionEvent a){
		if(count-- == 0){			
			timer.stop(); 
			removeRat(); 
			display.tempRats.setVisible(false);
			timer = null;		
		}
		
		display.tempRats.setText("" + count);
	}
	
	public void restart(){		
		position.removeFigure();
		setRandomRat();
	}
	
	public void removeRat(){
		position.removeFigure();
		setRandomRat();		
	}
	
	public void eatRat(){
		if(timer != null){
			timer.stop();			
			display.tempRats.setVisible(false);
			timer = null;			
		}
		
		setRandomRat();		
	}
	
	private class RatFigure extends Figure{
		 public RatFigure() { super(); }
		 private final Color BROWN = new Color(128, 64, 0);
		 
		 public void draw(Graphics canvas, int x, int y){
			if(timer == null)
			{canvas.setColor(BROWN);}
			else{canvas.setColor(Color.yellow);}
			
			canvas.fillOval(0, 0, x, y);		
			canvas.setColor(Color.BLACK);
			canvas.drawOval(0, 0, x, y);
		 }
	}
}

File: src/main/kotlin/snake/models/Rat.kt
package snake.models

import snake.game.GameDataDisplay
import snake.models.basic.Point
import snake.views.RatView
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.util.*
import javax.swing.Timer

class Rat(private val board: GameMatrix, private val display: GameDataDisplay) : ActionListener {
    var position: Point? = null
        private set
    private val timer: Timer = Timer(SECOND, this)
    private val random: Random = Random()
    var numberOfRats = 0
        private set
    private var count = 0

    private fun setRandomRat() {
        var x: Int
        var y: Int
        do {
            x = random.nextInt(board.width)
            y = random.nextInt(board.height)
        } while (!board.isEmpty(x, y))
        if (random.nextInt(100) > 90) {
            count = DISAPPEARING_SECONDS
            numberOfRats = random.nextInt(5) + 1
            display.tempRats.text = "" + count
            display.tempRats.isVisible = true
            timer.start()
        }
        val position = Point(x, y)
        board.setView(RatView(isTimed), position)
        this.position = position
    }

    val isTimed: Boolean = timer.isRunning

    override fun actionPerformed(a: ActionEvent) {
        if (count-- == 0) {
            timer.stop()
            removeRat()
            display.tempRats.isVisible = false
        }
        display.tempRats.text = "" + count
    }

    fun restart() {
        board.removeView(position!!)
        setRandomRat()
    }

    fun removeRat() {
        board.removeView(position!!)
        setRandomRat()
    }

    fun eatRat() {
        if (timer.isRunning) {
            timer.stop()
            display.tempRats.isVisible = false
        }
        setRandomRat()
    }

    fun inPosition(point: Point?): Boolean {
        return position == point
    }

    companion object {
        private const val SECOND = 1000
        const val DISAPPEARING_SECONDS = 10
    }

    init {
        setRandomRat()
    }
}